### ISSUE TITLE
Prototype snapshot defragmentation

### ISSUE THREAD

**Issue by npepinpe** [camunda-cloud/zeebe#6930]
**Description**

It seems possible and cost-effective to defragment a highly fragmented RocksDB snapshot (i.e. > 100k small SST files) by simply creating a new, empty database, and copying key-by-key from the old database into the new one. Since high fragmentation doesn't seem related to the actual size of the database (as in, number of bytes contained in it), it can be much more efficient to open it and copy each key then actually copy the database itself - and defragmenting also has many additional benefits (faster checksum computation, faster recovery, faster replication, etc.).

While the proposal of copying is simple/easy enough, snapshots are critical to maintain the integrity of the log, and as such this is an operation which needs to be executed safely (and rolled back safely on any error), which is why I would like to first do a quick prototype to understand what has to be done. This will allow us to evaluate if we should commit resources to this, as we expect snapshot fragmentation to be very low with 1.0 (though not 100% proven) - but we are still supporting 0.26 until 1.1, and as there's no migration path from 0.26 to 1.0, it's unclear how many users and how fast they will upgrade, so it might be worth investing time in this.

This is to be a follow up of #6929.

**Comment by deepthidevaki** [comment]
I tried a different approach to compact the state. Using forced manual compaction (https://github.com/facebook/rocksdb/wiki/Manual-Compaction), I was able to reduce the number of files in a snapshot from 1065 to 24.

First I tried to compact using `compactRange` method. But it did not compact anything. But `compactFiles` worked.

What I did so far is to force compaction during leader transition. Immediately after zbdb is opened (after copying the snapshot to the runtime), a manual compaction is enforced. Only after the compaction is done stream processor is started. 

See branch `dd-force-rocksb-compaction`.
https://github.com/camunda-cloud/zeebe/blob/8b97ad468f4ff16d478e72d10bab7d84e607a329/zb-db/src/main/java/io/zeebe/db/impl/rocksdb/transaction/ZeebeTransactionDb.java#L192-L210

**Comment by npepinpe** [comment]
Nice! I had tried compact range but as you said, nothing much happened. I wasn't aware of the other API, good catch! I have some questions, but I think this is the best approach to take imo.

Do we know how long this could take? Could it be prohibitively expensive during leader transition?

**Comment by deepthidevaki** [comment]
> Do we know how long this could take? Could it be prohibitively expensive during leader transition?

It would be easy to measure how long it takes. Alternatively, we can trigger this only on specific conditions.

**Comment by deepthidevaki** [comment]
Around 300ms for 1000 files.
```
I 2021-05-18T12:33:50.843736Z ATTEMPTED to compact on startup - file count before compation = 1020 after compaction = 28. Compaction took 258ms 
I 2021-05-18T12:33:52.715086Z ATTEMPTED to compact on startup - file count before compation = 1018 after compaction = 35. Compaction took 323ms 
I 2021-05-18T12:33:54.157456Z ATTEMPTED to compact on startup - file count before compation = 943 after compaction = 27. Compaction took 264ms 
```

**Comment by npepinpe** [comment]
:+1: 

Very nice! I would propose we add it - let's add a new configuration parameter which can serve as a toggle/feature flag to enable this, with it being on by default, just to be safe. If we plan to forward port this to 1.0, then I would propose to add a metric, but I'm not sure this is useful for us in 1.0 (unless we again look into having multiple column families in the future, but that's unclear as well).

**Comment by npepinpe** [comment]
From my POV the issue can be closed as successful - this came from https://github.com/camunda-cloud/zeebe/issues/6854, so the solution can target that issue (and there may be more?). I think your solution is really good, and is simple enough that we can roll it out to existing users in the next patch release (I would like to release one early next week, let me know if this would not be possible).

**Comment by deepthidevaki** [comment]
There are some more questions to be answered to ensure that it is safe to do this. 

1. I have now specified the output level to be 2. Is that a sane value? What would happen if there are files in higher levels - would they be compacted or ignored? 

2. I also saw some exceptions when trying to compact a snapshot with several 1000s of files. Why does this happen? Is it safe to ignore this exception? Eventhough there was this exception, it did manage to compact and the number of files reduced to almost half.
```
org.rocksdb.RocksDBException: Necessary compaction input file /000905.sst is currently being compacted.
        at org.rocksdb.RocksDB.compactFiles(Native Method)
        at org.rocksdb.RocksDB.compactFiles(RocksDB.java:3577)
        at io.zeebe.db.impl.rocksdb.transaction.ZeebeTransactionDb.compactFiles(ZeebeTransactionDb.java:203)
```

**Comment by deepthidevaki** [comment]
Superseded by #7086
